### DISCUSSION TITLE
RFC: Allow for each layer to be a different shape

### DISCUSSION THREAD

**Discussion by TimTheBig** [GladiusSlicer/GladiusSlicer#20] [Ideas]
To achieve this the concept of a layer needs to be separated from a flat plane. Possibly using some kind of mesh that shapes it's self to the model.
This would allow for features like non planar layers, and conical slicing. This would be an internal feature to enable other user facing features.

**Comment by TimTheBig** [comment]
I'm looking to do some thing based on [CurviSlicer](https://hal.science/hal-02120033/document) but directly in the slicer for integration with other slicer features and easy use.

**Reply by TimTheBig** [reply]
[Radial Non Planar Slicer](https://github.com/jyjblrd/Radial_Non_Planar_Slicer) is a new method worth looking in to.

**Comment by mrhatman** [comment]
This could be a fairly large change. I would want to spend time architecting the changes to make sure that the changes don't make the code difficult to maintain ( not that its in a great place right now).

A good first step would be something like the CR-30.

**Reply by TimTheBig** [reply]
**Small steps** are the way to go for some thing this big, I don't have a belt printer to test with, but adding support will be a good way to slowly introduce flexibility to the slicer.

**Reply by TimTheBig** [reply]
Sample CR-30 g-code:
[CBelt45_04_PLA_Voron_Design_Cube_v7.gcode.zip](https://github.com/user-attachments/files/17543422/CBelt45_04_PLA_Voron_Design_Cube_v7.gcode.zip)

**Comment by TimTheBig** [comment]
This [paper](https://tams.informatik.uni-hamburg.de/publications/2018/MSc_Daniel_Ahlers.pdf) may be closer to what I'm going for.

**Comment by TimTheBig** [comment]
What files should I look at for adding belt printer support?

**Comment by mrhatman** [comment]
Fundamentally the changes boil down to:

- Changes to the slicing portion ( mostly in tower.rs) to slicer at different angle
    - The tower uses a sweeping plane algorithm for slicing. Assuming the sweep is still planar, all that needs to change is that the sorting function changes
    - Planar here can be defined as, for each triangle the sweep plane will hit a vertex of the triangle before it hits a edge or the face. A lot of assumptions about geometry fall apart if that doesn't hold true.
- The ability to customize the passes for the geometry
    - for example the overhang logic is different for a belt printer
- changes to the plotter to convert the flat 2d representation into the final geometry
    - not much of a difference for a belt printer as the final print is still flat. But if we wanted to test this on a normal printer, then each line needs to be drawn in 3 dimensions instead of 2.

**Reply by TimTheBig** [reply]
Here is a draft sorting function for  tower.rs:
```rs
// Sort tower vertices based on their projection along the normal vector
tower_vertices.sort_by(|a, b| {
    let proj_a = vertices[a.start_index].dot(&normal); // Calculate projection for vertex a
    let proj_b = vertices[b.start_index].dot(&normal); // Calculate projection for vertex b
    proj_a
        .partial_cmp(&proj_b)
        .expect("STL ERROR: No Points should have NAN values")
});
```

```rs
impl Vertex {
    fn dot(&self, other: &Vertex) -> f64 {
        self.x * other.x + self.y * other.y + self.z * other.z
    }
}
```

**Comment by chungwong** [comment]
What is [TriangleTower](https://github.com/GladiusSlicer/GladiusSlicer/blob/main/src/tower.rs#L46)?
I don't have a mental visualisation for TriangleTower. How does it relate to a 3D model?

What is the purpose of this transformation(3d Model => Tower)?

**Reply by mrhatman** [reply]
Triangle Tower is just a data structure that allows for slicing. 

The Tower stores the transformed vertices and a reference the the vertices and faces that are connected to that vertices that lie "above" it.

Then the Tower iterator allows will perform a sweeping plane algorithm to slice into each layer.

Importantly for this discussion, the ordering can be changed to effect what is above and the sweeping plane does not have to be flat.

**Reply by chungwong** [reply]
How did you come up with the algorithm? Did you reference any research papers?
I am also trying to understand(visually) the steps/logic for the tower and also slicing. At the moment I am trying to visualize it using [Bevy](https://github.com/bevyengine/bevy)

**Reply by mrhatman** [reply]
This was derived from a class I took in College on computational geometry. Its based on some sweep line algorithms like [Forunes](https://en.wikipedia.org/wiki/Fortune%27s_algorithm) or [Bentley Ottmann](https://en.wikipedia.org/wiki/Bentley%E2%80%93Ottmann_algorithm)

The algorithm is terribly documented. I just opened #26 to track this progress.

**Comment by TimTheBig** [comment]
I have spent a month now working on this. I can't get past the tower rings being incomplete on the first layer and all subsequent layers.

```rs
2024-11-25T14:18:44.242Z ERROR [gladius_slicer::utils] **************************************************
2024-11-25T14:18:44.242Z ERROR [gladius_slicer::utils]  Gladius Slicer Ran into an error
2024-11-25T14:18:44.242Z ERROR [gladius_slicer::utils]  Error Code: 0x1007
2024-11-25T14:18:44.242Z ERROR [gladius_slicer::utils]  Error Creating Tower. Model most likely needs repair. Please Repair and run again: Incomplete ring detected at height 102.176
2024-11-25T14:18:44.242Z ERROR [gladius_slicer::utils] **************************************************
2024-11-25T14:18:44.242Z ERROR [gladius_slicer::utils] 
```

**Reply by mrhatman** [reply]
Where is your code? I can take a look.

**Reply by TimTheBig** [reply]
[This repo](https://github.com/TimTheBig/GladiusSlicer/tree/angled-slice)

**Reply by mrhatman** [reply]
Have you tried with a completely level plane?

The issue is likely in from_triangles_and_vertices. The points are ordered there as well. 

The implementation should match your new sort function

**Reply by TimTheBig** [reply]
Even with a level plane the issue still occurs.

**Reply by TimTheBig** [reply]
> The implementation should match your new sort function
What do you mean be that?

**Reply by mrhatman** [reply]
Your new sort function orders in reference to the normal.

Line 70-71
`
 // depending what is the next vertex is its either leading or trailing
            if vertices[index_tri.verts[1]] < vertices[index_tri.verts[2]] {
`

this is based on the Ord implimentation for vertex.

These need to match for the everything to function.

**Reply by TimTheBig** [reply]
I tried this how ever it's still running in to the same problem:
```rs
if vertices[index_tri.verts[1]].dot(normal) < vertices[index_tri.verts[2]].dot(normal)
```
Now the incomplete rings are too short.
